### PR TITLE
helm: support common envFrom for Ray containers

### LINKED ISSUE #5021: [Feature] Native support for injecting Secrets via Helm chart values
### Search before asking

- [x] I had searched in the [issues](https://github.com/ray-project/kuberay/issues) and found no similar feature requirement.

### KubeRay Component

ray-operator

### Description

Today, there's no built-in way to inject sensitive values (private PyPI/pip index credentials, API tokens, cloud credentials, etc.) into Ray head or worker pods through the chart's `values.yaml`.

Because `headGroupSpec` and `workerGroupSpecs` don't expose a way to reference Kubernetes Secrets, the only option today is to patch the rendered manifests with Kustomize to add `env`, `secretKeyRef`, or Secret-backed volumes.

While this works, it adds unnecessary complexity for what feels like a common deployment requirement. It also has a few drawbacks:

- Requires an additional Kustomize layer on top of Helm for a basic configuration need.
- Isn't very discoverable—neither `values.yaml` nor the documentation points users toward this workflow.
- The same patch has to be applied to `headGroupSpec` and every `workerGroupSpec`, which becomes repetitive as deployments grow.
- There's no single place in `values.yaml` to reference Secrets (including those managed by External Secrets Operator) and have them applied consistently across all Ray pods.

### Use case

Many teams need to provide sensitive configuration such as private PyPI/pip index credentials, application API tokens, cloud credentials, or object storage credentials without baking them into container images or committing them as plaintext.

It would be helpful if the chart provided a first-class way to declare Secret-backed environment variables and/or volume mounts once in `values.yaml`, with those settings automatically applied to both head and worker pod templates.

This would:

- Remove the need for Kustomize overlays for a common day-one deployment requirement.
- Make secret configuration easier to discover and maintain.
- Provide a single, consistent place to configure Secrets for all Ray pods.
- Better support common Kubernetes workflows, including Secrets managed by External Secrets Operator.

### Related issues

didnt find any related to it. 

### Are you willing to submit a PR?

- [ ] Yes I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by Smallfu666**
## Why are these changes needed?

The `ray-cluster` chart already exposes `envFrom` on the head, the default worker, and each `additionalWorkerGroups` entry, so Secret-backed environments do work today — but there is no single place to apply the same sources to every Ray container. As #5021 points out, that means repeating the same block for each group, which is what pushes people toward a Kustomize layer for something the chart can already almost express.

This adds `common.envFrom`, following the existing `common.containerEnv` pattern, so a Secret managed by an external controller (e.g. External Secrets Operator) can be wired in once:

```yaml
common:
  envFrom:
    - secretRef:
        name: my-env-secret
```

How it behaves:

- Common sources are rendered **before** each group's local `envFrom` sources. Since Kubernetes lets a later source win for a duplicate key, group-local values keep precedence.
- All four render paths are covered: head, the default `worker`, the built-in `smallGroup`, and custom `additionalWorkerGroups` entries — including groups that define no local `envFrom` at all.
- When both lists are empty, no `envFrom` field is rendered, preserving the existing default behavior.

This PR intentionally limits the shared chart surface to `envFrom`; ESO resources, shared volumes, sidecars, init containers, and autoscaler configuration remain unchanged.

## Related issue number

Closes #5021

## Checks

- [x] I've made sure the tests are passing.
- [x] Testing Strategy
  - [x] Unit tests: four new cases in `helm-chart/ray-cluster/tests/raycluster_test.yaml` — ordering asserted against the full list on head / `worker` / `smallGroup`; `common.envFrom` applied to a custom worker group with no local `envFrom` (the nil-safe path); `notExists` on both head and worker when the lists are empty; and duplicate entries preserved when `common` and a group reference the same source.

    ```
    helm unittest helm-chart/ray-cluster --file "tests/**/*_test.yaml" --strict
    # Tests: 108 passed, 108 total

    helm lint helm-chart/ray-cluster
    # 1 chart(s) linted, 0 chart(s) failed
    ```
  - [x] `make -C helm-chart helm-docs` regenerated the chart README and is reproducible on rerun.
  - [x] Rendered all four paths manually and confirmed only the `RayCluster` object is produced.

This contribution was developed with AI assistance and reviewed and tested end-to-end by the submitter.

### CODE REVIEW

File: helm-chart/ray-cluster/values.yaml [hunk 1]
```diff
@@ -56,6 +56,11 @@ common:
   containerEnv: []
   #  - name: BLAH
   #    value: VAL
+
+  # -- envFrom specifies sources used to populate environment variables in the main Ray head and worker containers. Common sources are rendered before group-specific sources. Referenced ConfigMaps and Secrets must already exist.
```
Comment (machichima) [inline helm-chart/ray-cluster/values.yaml]:
```suggestion
  # -- envFrom specifies sources used to populate environment variables in the main Ray head and worker containers.
  # Common sources are rendered before group-specific sources. Referenced ConfigMaps and Secrets must already exist.
```

**Comment by Smallfu666** [comment]
Thanks @machichima — applied in `6ac1cc1`. Confirmed `make -C helm-chart helm-docs` produces no README change (helm-docs joins the `# --` comment lines), so the rendered table row is unchanged.

**Review by machichima — APPROVED** [review]
LGTM! Thank you.
cc @rueian @andrewsykim to merge

File: helm-chart/ray-cluster/tests/raycluster_test.yaml [hunk 2]
```diff
@@ -2240,3 +2240,84 @@ tests:
...
+        customGroup:
+          replicas: 1
+    asserts:
+      - equal:
+          path: spec.workerGroupSpecs[?(@.groupName=="customGroup")].template.spec.containers[?(@.name=="ray-worker")].envFrom
+          value:
+            - secretRef:
+                name: common-secret
+
+  - it: Should not render envFrom when both common and local lists are empty
```
Comment (andrewsykim) [inline helm-chart/ray-cluster/tests/raycluster_test.yaml:2318]: Should we add a test where both local and common list have same env var? Is that supported or will it result in duplicate entries?

**Review by andrewsykim — APPROVED** [review]

File: helm-chart/ray-cluster/tests/raycluster_test.yaml (hunk 2 above)
Comment (machichima) [reply helm-chart/ray-cluster/tests/raycluster_test.yaml:2318]: I think it will cause duplicate entries and the latter one will be used. Agree we can add a test case for this

Comment (Smallfu666) [reply helm-chart/ray-cluster/tests/raycluster_test.yaml:2318]: Added the test, thanks!

File: helm-chart/ray-cluster/tests/raycluster_test.yaml [hunk 3]
```diff
@@ -2240,3 +2240,114 @@ tests:
...
+  - it: Should keep duplicate entries when common and group envFrom reference the same source
+    set:
+      common:
+        envFrom:
+          - configMapRef:
+              name: shared-config
+      head:
+        envFrom:
+          - configMapRef:
+              name: shared-config
```
Comment (machichima) [inline helm-chart/ray-cluster/tests/raycluster_test.yaml]:
```suggestion
              name: shared-config-head
```
nit: use different value as in `common`

File: helm-chart/ray-cluster/tests/raycluster_test.yaml [hunk 4]
```diff
@@ -2240,3 +2240,114 @@ tests:
...
+      worker:
+        envFrom:
+          - configMapRef:
+              name: shared-config
```
Comment (machichima) [inline helm-chart/ray-cluster/tests/raycluster_test.yaml]:
```suggestion
              name: shared-config-worker
```
ditto

**Review by machichima — APPROVED** [review]
LGTM! Thank you. Just nit

File: helm-chart/ray-cluster/tests/raycluster_test.yaml (hunk 3 above)
Comment (Smallfu666) [reply helm-chart/ray-cluster/tests/raycluster_test.yaml]: Updated. Thanks

File: helm-chart/ray-cluster/tests/raycluster_test.yaml (hunk 4 above)
Comment (Smallfu666) [reply helm-chart/ray-cluster/tests/raycluster_test.yaml]: Updated. Thanks

**Review by rueian — APPROVED** [review]
